integrations/micronaut/cdi/src/main/java/io/helidon/integrations/micronaut/cdi/MicronautCdiExtension.java

```diff
@@ -0,0 +1,405 @@
+
+                        Set<Class<? extends MethodInterceptor<?, ?>>> interceptors = new HashSet<>();
+                        methodInterceptors.forEach(it -> interceptors.add((Class<? extends MethodInterceptor<?, ?>>) it));
+
+                        methods.computeIfAbsent(javaMethod,
+                                                theMethod -> MethodInterceptorMetadata.create(
+                                                        method.getAnnotated(),
+                                                        executableMethodCache.get(theMethod)))
+                                .addInterceptors(interceptors);
+                    }
+                });
+    }
+
+    /**
+     * Add all (not yet added) Micronaut beans for injection as long as they are singletons.
+     *
+     * @param event CDI event
+     */
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void afterBeanDiscovery(@Priority(PLATFORM_BEFORE) @Observes AfterBeanDiscovery event) {
+        event.addBean()
+                .addType(ApplicationContext.class)
+                .id("micronaut-context")
+                .scope(ApplicationScoped.class)
+                .produceWith(instance -> micronautContext.get());
+
+        // add the remaining Micronaut beans
+        for (var entry : unprocessedBeans.entrySet()) {
+            Class<?> beanType = entry.getKey();
+            List<MicronautBean> beans = entry.getValue();
+
+            List<? extends BeanDefinitionReference<?>> refs = List.of();
+            if (beans.size() > 1) {
+                // first make sure these are singletons; if not, ignore
+                refs = beans.stream()
+                        .map(MicronautBean::definitionRef)
+                        .filter(it -> !it.getBeanType().getName().endsWith("$Intercepted"))
+                        .filter(BeanDefinitionReference::isSingleton)
+                        .collect(Collectors.toList());
+            }
+
+            // primary
+            event.addBean()
+                    .addType(beanType)
+                    .id("micronaut-" + beanType.getName())
+                    // inject using dependent - manage scope by micronaut context
+                    .scope(Dependent.class)
+                    .produceWith(instance -> micronautContext.get().getBean(beanType));
+
+            if (refs.size() > 1) {
+                // we must care about qualifiers
+                for (var ref : refs) {
+                    AnnotationMetadata annotationMetadata = ref.getAnnotationMetadata();
+                    List<Class<? extends Annotation>> qualifiers = annotationMetadata
+                            .getAnnotationTypesByStereotype(Qualifier.class);
+
+                    Annotation[] synthesized = new Annotation[qualifiers.size()];
+                    for (int i = 0; i < qualifiers.size(); i++) {
+                        synthesized[i] = annotationMetadata.synthesize(qualifiers.get(i));
```

Comment:
I think you can probably create the `io.micronaut.context.Qualifier[]` in the same loop with `Qualifiers.byAnnotation(annotationMetadata, qualifiers.get(i))`